### PR TITLE
fix: percentile operator bug in limit queries

### PR DESCRIPTION
**Pull request by makeavish**
close(https://github.com/SigNoz/signoz/issues/3204)

### CODE REVIEW

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 1]
```diff
@@ -228,7 +228,7 @@ func buildLogsQuery(panelType v3.PanelType, start, end, step int64, mq *v3.Build
 	}
 
 	if graphLimitQtype == constants.SecondQueryGraphLimit {
-		filterSubQuery = filterSubQuery + " AND " + fmt.Sprintf("(%s) GLOBAL IN (", getSelectKeys(mq.AggregateOperator, mq.GroupBy)) + "%s)"
+		filterSubQuery = filterSubQuery + " AND " + fmt.Sprintf("(%s) GLOBAL IN (", getSelectKeys(mq.AggregateOperator, mq.GroupBy)) + "{{.FilterQuery}})"
```
Comment (nityanandagohain) [inline pkg/query-service/app/logs/v3/query_builder.go:231]: FilterSubquery

File: pkg/query-service/app/queryBuilder/query_builder.go [hunk 2]
```diff
@@ -201,7 +211,15 @@ func (qb *QueryBuilder) PrepareQueries(params *v3.QueryRangeParamsV3, args ...in
...
 							return nil, err
 						}
-						query := fmt.Sprintf(placeholderQuery, limitQuery)
+						data := map[string]string{
+							"FilterQuery": limitQuery,
+						}
+						t := template.Must(template.New("query").Parse(placeholderQuery))
+						builder := &strings.Builder{}
+						if err := t.Execute(builder, data); err != nil {
+							panic(err)
```
Comment (nityanandagohain) [inline pkg/query-service/app/queryBuilder/query_builder.go:220]: return err

File: pkg/query-service/app/queryBuilder/query_builder.go [hunk 3]
```diff
@@ -181,7 +182,16 @@ func (qb *QueryBuilder) PrepareQueries(params *v3.QueryRangeParamsV3, args ...in
...
```
Comment (nityanandagohain) [inline pkg/query-service/app/queryBuilder/query_builder.go:191]: return err

**Review by nityanandagohain — APPROVED** [review]

**Comment by srikanthccv** [comment]
Please help me understand why templating is needed here

**Comment by makeavish** [comment]
> Please help me understand why templating is needed here

If user uses `%` in the filter query then `fmt.Sprintf` was failing as `%` was treated as variable insertion operator. We could have escaped the characters but that could have more code changes. So used templates.

**Comment by srikanthccv** [comment]
I do not think templating should be used because we need to make more code changes. What more changes are we talking about? There is one `ClickHouseFormattedValue` which should take of it.

**Comment by makeavish** [comment]
> I do not think templating should be used because we need to make more code changes. What more changes are we talking about? There is one `ClickHouseFormattedValue` which should take of it.

Apart from filter `%` can occur at multiple places, what if `%` is present in a key? So the cases keeps on increasing. 
Using Template was safest bet for future changes or such scenarios.

**Comment by srikanthccv** [comment]
>Apart from filter % can occur at multiple places, what if % is present in a key? So the cases keeps on increasing

This is very unlikely. I don't think that justifies it. Assume such a key exists for the sake of argument. Does this fix solve all the issues? This is simply checking specific limit cases.

**Comment by makeavish** [comment]
> > Apart from filter % can occur at multiple places, what if % is present in a key? So the cases keeps on increasing
> 
> This is very unlikely. I don't think that justifies it. Assume such a key exists for the sake of argument. Does this fix solve all the issues? This is simply checking specific limit cases.

In cases where we are using limit, it won't be a reproducible but template doesn't fix other such possible cases. We can use template everywhere to avoid the issues.

**Comment by srikanthccv** [comment]
My point is that issue arose because of a special character, and it should be escaped. The idea of templatizing the whole query builder is a new effort from scratch; having seen the builder code, I think it will turn out to be a net negative and (most likely) will not fix the special character problems. I do not think we should introduce a templating here, and "mode code changes" is not a justifiable reason. I can think of many cases where not escaping is the source of bugs (try `level = 'info` in logs or something similar in traces or metrics). I do not want this PR to fix all potential such issues but escaping is the way to go, and it should start with this. I am open here others' thoughts on this.

// @nityanandagohain @rkssisodiya @ankitnayan

**Comment by ankitnayan** [comment]
There must be many golang projects that deal with the same issue. It would be wise to look into a few of them and get inspiration. @makeavish please post your findings

**Comment by makeavish** [comment]
Other projects uses DB library query parameters methods instead of manual escaping. We should also use clickhouse named parameters. We are already using clickhouse named parameters for `getFilteredSpan` function in `reader.go`.

**Comment by makeavish** [comment]
Closing this as we will go ahead with using clickhouse named operators: https://github.com/SigNoz/signoz/issues/3345

**Pull request closed without merging**